test/agent.js

```diff
@@ -1,533 +1,486 @@
+//     res.end(wanted)
+//   })
+
+//   t.tearDown(server.close.bind(server))
+
+//   server.listen(0, () => {
+//     const agent = new Agent()
+
+//     const origin = `http://localhost:${server.address().port}`
+
+//     request(origin, { agent })
+//       .then(() => {
+//         t.fail()
+//       })
+//       .catch(err => {
+//         t.ok(err instanceof errors.ClientDestroyedError)
+//       })
+
+//     const pool = agent.get(origin)
+//     pool.once('connect', () => {
+//       agent.destroy().then(() => {
+//         request(origin, { agent })
+//           .then(() => {
+//             t.fail()
+//           })
+//           .catch(err => {
+//             t.ok(err instanceof errors.ClientDestroyedError)
+//           })
+//       })
+//     })
+//   })
+// })
+
+test('multiple connections', t => {
+  const connections = 3
+  t.plan(6 * connections)
+
+  const server = http.createServer((req, res) => {
+    res.writeHead(200, {
+      Connection: 'keep-alive',
+      'Keep-Alive': 'timeout=1s'
     })
-    t.test('agent should destroy internal pools', t => {
-      t.plan(2)
+    res.end('ok')
+  })
+  t.tearDown(server.close.bind(server))
 
-      const wanted = 'payload'
+  server.listen(0, async () => {
+    const origin = `http://localhost:${server.address().port}`
+    const agent = new Agent({ connections })
 
-      const server = http.createServer((req, res) => {
-        res.setHeader('Content-Type', 'text/plain')
-        res.end(wanted)
-      })
+    t.tearDown(agent.close.bind(agent))
 
-      t.tearDown(server.close.bind(server))
-
-      server.listen(0, () => {
-        const agent = new Agent()
-
-        const origin = `http://localhost:${server.address().port}`
-
-        request(origin, { agent })
-          .then(() => {
-            t.fail()
-          })
-          .catch(err => {
-            t.ok(err instanceof errors.ClientDestroyedError)
-          })
-
-        const pool = agent.get(origin)
-        pool.once('connect', () => {
-          agent.destroy().then(() => {
-            request(origin, { agent })
-              .then(() => {
-                t.fail()
-              })
-              .catch(err => {
-                t.ok(err instanceof errors.ClientDestroyedError)
-              })
-          })
-        })
-      })
+    agent.on('connect', (client) => {
+      t.ok(client)
+    })
+    agent.on('disconnect', (client, error) => {
+      t.ok(client)
+      t.true(error instanceof errors.InformationalError)
+      t.strictEqual(error.code, 'UND_ERR_INFO')
+      t.strictEqual(error.message, 'reset')
     })
+
+    for (let i = 0; i < connections; i++) {
+      await request(origin, { agent })
+        .then(() => {
+          t.pass('should pass')
+        })
+        .catch(err => {
+          t.fail(err)
+        })
+    }
   })
+})
 
-  t.test('Agent connect/disconnect event(s)', t => {
-    t.plan(2)
+// t.test('remove disconnect listeners when destroyed', t => {
```

Comment:
@ronag In this test what fails is the check after calling `Agent.get(origin)`. The previous implementation attached the event listeners when calling `get`:

https://github.com/nodejs/undici/blob/main/lib/agent.js#L52

Reply:
Not sure I follow... nothing has changed there in this PR?

Reply:
Sorted. Thanks!